### PR TITLE
Bug 2053685: (Topology) Performance improvement by reducing rerenderings and deep-copy toJSON() calls

### PR DESCRIPTION
**Pull request by jerolimov**
**Fixes**: 
https://bugzilla.redhat.com/show_bug.cgi?id=2053685

**Tl;dr**
First commit adds only tests, second commit adds caching for previously deep-cloned data.
This cache of the JSON data which are already used in the topology allows the topology and many other components to do less useMemo, useEffect, re-calculations and re-renderings. In topology, this happens because an object replacement in a GraphNode rerenders it.

**Analysis / Root cause**:
While analyzing our topology performance issues I found two (main) issues which are addressed in this PR. (This is not the first and not the last PR for topology.)

1. The topology components rerenders too much, also when only a part of the data has changed.
2. A lot of time is consumed in the immutable `toJSON()` method.

After further investigation I noticed that the toJSON function is called in `Firehose`, `useK8sWatchResource` and `useK8sWatchResources` when converting an immutable object to JSON. This happens whenever a component is rendered and uses one of the hooks.

It also rerenders or recalculates hooks to often with the same data because the created JSON object contains the same data in a new, deep-clone object or array. But `Memo`d-components, hooks dependencies, and other optimizations like the topology `mobx` graph tree depends on an identity to check to rerender components..

**Solution Description**: 
This PR improves the topology performance or the performance of many different pages with an "immutable toJSON" result cache in `firehose.jsx` and `k8s-watcher.tsx`

It adds a cache and saves the converted JSON data at the immutable object itself. (Thanks Christian for this great idea.)

:recycle:  To see and test what will change with this caching I first added some tests to the Firehose component and the hooks.

1. The first commit just adds test for the status quo.
2. The second commit adds the caching and updates the tests so that it should be easier to see which objects are reused now.

:question: Why not just remove Immutable from the Redux store?!

I think removing Immutable would be a good idea. For the future.

But adding this cache has some benefits for the moment: It's easier and safer to add/test/verify and I hope that we can backport this. I don't expect that we would backport a rewrite of the reducers, and selectors, etc.

:warning:  Potential risk: Reused data models (redux state) is now reused over time and by different components. Changing such a data object would be a bad practice similar to changing a props object.

#### Before - Load topology with many deployments (in my case 84)

https://user-images.githubusercontent.com/139310/152451159-4b2612b5-0391-435e-be42-97dfd986c369.mp4

#### After - Load topology with many deployments (in my case 84)

https://user-images.githubusercontent.com/139310/152451174-76b9dcfc-4cbc-4297-b57c-1ee75c7e80fe.mp4

#### Before - Stay on the topology page with many deployments (here 84) and delete a random pod every 3 second

https://user-images.githubusercontent.com/139310/152451186-cc01c17a-d63b-4f42-9bab-943584a100c3.mp4

#### After - Stay on the topology page with many deployments (here 84) and delete a random pod every 3 second

https://user-images.githubusercontent.com/139310/152451200-16fd63f9-74d2-437a-a917-2c37264d235e.mp4

#### Before - Load topology with many deployments (in my case 84)

before-topology-first-40sec

#### After - Load topology with many deployments (in my case 84)

after-topology-load

#### Before - Stay on the topology page with many deployments (here 84) and delete a random pod every 3 second

before-topology-updatepods-40sec

#### After - Stay on the topology page with many deployments (here 84) and delete a random pod every 3 second

todo

**Screen shots / Gifs for design review**: 
UI isn't changed. Here are som

**Unit test coverage report**: 
Added a lot of tests for the existing components `Firehose`, `useK8sWatchResource` and `useK8sWatchResources` to check the status quo and track the difference with this PR.

**Test setup:**
1. Create a project/namespace with a lot of Deployments (50+)
2. Open the topology
3. Delete some pods

You can find some scripts to create a cluster with a lot of load here: https://github.com/jerolimov/openshift/tree/master/loadtest

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts [hunk 1]
```diff
@@ -47,15 +47,25 @@ export const makeQuery: MakeQuery = (namespace, labelSelector, fieldSelector, na
...
+    return immutableData.toArray().map((a) => {
+      if (!a[CACHE_SYMBOL]) {
+        a[CACHE_SYMBOL] = a.toJSON();
+      }
+      return a[CACHE_SYMBOL];
+    });
   }
-  if (immutableData.toJSON) {
-    return immutableData.toJSON();
+  if (immutableData.toJS) {
```
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts:64]: Shouldn't this still be:
```suggestion
  if (immutableData.toJSON) {
```

File: frontend/public/components/utils/firehose.jsx [hunk 2]
```diff
@@ -19,7 +19,9 @@ const shallowMapEquals = (a, b) => {
   return a.every((v, k) => b.get(k) === v);
 };
 
-const processReduxId = ({ k8s }, props) => {
+const CACHE_SYMBOL = Symbol('_cachedToJSResult');
```
Comment (christianvogt) [inline frontend/public/components/utils/firehose.jsx:22]: Safe to use the same symbol as the watcher. However since `firehose` is old and would probably be more grief to share the symbol with the sdk.

**Comment by christianvogt** [comment]
@jerolimov I agree that removing immutable JS is something we should look at doing but it's a much larger change. This is a good first step and proves there's a larger issue to address with immutable js and our redux store.

File: frontend/public/components/utils/firehose.jsx (hunk 2 above)
Comment (jerolimov) [reply frontend/public/components/utils/firehose.jsx:22]: I added this to the SDK because we have already an import from firehose.jsx to the SDK. I'm not sure what the "right' import direction here....?!

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/__tests__/useK8sWatchResources.data.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    apiVersion: 'v1',
+    kind: 'Pod',
+    metadata: {
+      name,
+      namespace: 'default',
+      resourceVersion: '123',
+    },
+  })),
+  metadata: { resourceVersion: '123' },
+};
```
Comment (vikram-raj) [inline frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/__tests__/useK8sWatchResources.data.tsx:26]: Looks like this file is duplicate of `frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/__tests__/useK8sWatchResource.data.tsx`.

File: frontend/public/components/utils/__tests__/firehose.data.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    apiVersion: 'v1',
+    kind: 'Pod',
+    metadata: {
+      name,
+      namespace: 'default',
+      resourceVersion: '123',
+    },
+  })),
+  metadata: { resourceVersion: '123' },
+};
```
Comment (vikram-raj) [inline frontend/public/components/utils/__tests__/firehose.data.tsx:28]: we can reuse `podList` and `podData` from `useK8sWatchResource.data.tsx`. I think it is better to use `podList` and `podData` from here and delete `useK8sWatchResource.data.tsx` and `useK8sWatchResources.data.tsx`. wdyt?

**Review by spadgett — APPROVED** [review]
/approve

I'm OK with the approach as a short-term fix we can backport. @jerolimov Let's open a follow up issue to remove immutable, which seems like a better long-term approach. We should try to give it priority since this change will increase the in-memory size of the k8s resource data.

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/__tests__/useK8sWatchResources.data.tsx (hunk 3 above)
Comment (jerolimov) [reply frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/__tests__/useK8sWatchResources.data.tsx:26]: Yeah I duplicated the test-data files so that they don't depend on each other.

File: frontend/public/components/utils/__tests__/firehose.data.tsx (hunk 4 above)
Comment (jerolimov) [reply frontend/public/components/utils/__tests__/firehose.data.tsx:28]: This is a different package. Yes, we have already dependencies from public package (console/internal) to  packages/console-dynamic-plugin-sdk but I would like to keep a copy here instead of adding more dependencies between both.

**Review by vikram-raj — COMMENTED** [review]
Verified it using cluster bot. In Chrome and Safari browsers with 100 D and 80 Pods in Topology and it is faster than before. 

/lgtm
